README.md
Pico Deprecated Plugin
======================

This is the repository of Pico's official `PicoDeprecated` plugin.

Pico is a stupidly simple, blazing fast, flat file CMS. See https://picocms.org/ for more info.

`PicoDeprecated`'s purpose is to maintain backward compatibility to older versions of Pico, by re-introducing characteristics that were removed from Pico's core. It for example triggers old events (like the `before_render` event used before Pico 1.0) and reads config files that were written in PHP (`config/config.php`, used before Pico 2.0).

Please refer to [`picocms/Pico`](https://github.com/picocms/Pico) to get info about how to contribute or getting help.

Install
-------

You usually don't have to install this plugin manually, it's shipped together with [Pico's pre-built release packages](https://github.com/picocms/Pico/releases/latest) and a default dependency of [`picocms/pico-composer`](https://github.com/picocms/pico-composer).

If you're using plugins and themes that are compatible with Pico's latest API version only, you can safely remove `PicoDeprecated` from your Pico installation or disable the plugin (please refer to the "Usage" section below). However, if you're not sure about this, simply leave it as it is - it won't hurt... :wink:

If you use a `composer`-based installation of Pico and want to either remove or install `PicoDeprecated`, simply open a shell on your server and navigate to Pico's install directory (e.g. `/var/www/html`). Run `composer remove picocms/pico-deprecated` to remove `PicoDeprecated`, or run `composer require picocms/pico-deprecated` (via [Packagist.org](https://packagist.org/packages/picocms/pico-deprecated)) to install `PicoDeprecated`.

If you rather use one of Pico's pre-built release packages, it is best to disable `PicoDeprecated` and not to actually remove it. The reason for this is, that `PicoDeprecated` is part of Pico's pre-built release packages, thus it will be automatically re-installed when updating Pico. However, if you really want to remove `PicoDeprecated`, simply delete the `plugins/PicoDeprecated` directory in Pico's install directory (e.g. `/var/www/html`). If you want to install `PicoDeprecated`, you must first create a empty `plugins/PicoDeprecated` directory on your server, [download the version of `PicoDeprecated`](https://github.com/picocms/pico-deprecated/releases) matching the version of your Pico installation and upload all containing files (esp. `PicoDeprecated.php` and the `lib/`, `plugins/` and `vendor/` directories) into said `plugins/PicoDeprecated` directory (resulting in `plugins/PicoDeprecated/PicoDeprecated.php`).

The versioning of `PicoDeprecated` strictly follows the version of Pico's core. You *must not* use a version of `PicoDeprecated` that doesn't match the version of Pico's core (e.g. PicoDeprecated 2.0.1 is *not compatible* with Pico 2.0.0). If you're using a `composer`-based installation of Pico, simply use a version constaint like `^2.0` - `PicoDeprecated` ensures that its version matches Pico's version. Even if you're using one of Pico's pre-built release packages, you don't have to take care of anything - a matching version of `PicoDeprecated` is part of Pico's pre-built release packages anyway.

Usage
-----

You can explicitly disable `PicoDeprecated` by adding `PicoDeprecated.enabled: false` to your `config/config.yml`. If you want to re-enable `PicoDeprecated`, simply remove this line from your `config/config.yml`. `PicoDeprecated` itself has no configuration options, it enables and disables all of its features depending on whether there are plugins and/or themes requiring said characteristics.

`PicoDeprecated`'s functionality is split into various so-called "compatibility plugins". There are compatibility plugins for every old API version (Pico 0.9 and earlier were using API version 0, Pico 1.0 was using API version 1 and Pico 2.0 was using API version 2; the current API version is version 3, used by Pico 2.1), one for plugins and another one for themes. Their purpose is to re-introduce characteristics plugins and themes using said API version might rely on. For example, plugin API compatibility plugins are responsible for simulating old Pico core events (like the `before_render` event used by Pico 0.9 and earlier). Theme API compatibility plugins will e.g. register old Twig variables (like the `is_front_page` Twig variable used by Pico 1.0). If you install a plugin using API version 2, the corresponding `picocms\PicoDeprecated\Plugin\PluginApi2Plugin` will be loaded. All plugin API compatibility plugins also depend on their theme counterpart, thus `picocms\PicoDeprecated\Plugin\ThemeApi2Plugin` will be loaded, too. Furthermore all compatibility plugins depend on their respective API successors.

The plugin exposes a simple API to allow other plugins to load their own compatibility plugins. As a plugin developer you may use the `PicoDeprecated::loadCompatPlugin(picocms\PicoDeprecated\PluginInterface $compatPlugin)` method to load a custom compatibility plugin. Use `PicoDeprecated::getCompatPlugins()` to return a list of all loaded compatibility plugins. You can furthermore use the `PicoDeprecated::getPlugins(int $apiVersion)` method to return a list of all loaded Pico plugins using a particular API version. If you want to trigger a custom event on plugins using a particular API version only, use `PicoDeprecated::triggerEvent(int $apiVersion, string $eventName, array $parameters = [])`. `PicoDeprecated` furthermore triggers the custom `onPicoDeprecated(PicoDeprecated $picoDeprecated)` event.

Getting Help
------------

Please refer to the ["Getting Help" section](https://github.com/picocms/Pico#getting-help) of our main repository.

Contributing
------------

Please refer to the ["Contributing" section](https://github.com/picocms/Pico#contributing) of our main repository.

By contributing to Pico, you accept and agree to the *Developer Certificate of Origin* for your present and future contributions submitted to Pico. Please refer to the ["Developer Certificate of Origin" section](https://github.com/picocms/Pico/blob/master/CONTRIBUTING.md#developer-certificate-of-origin) in the `CONTRIBUTING.md` of our main repository.


CHANGELOG.md
Pico Deprecated Changelog
=========================

**Note:** This changelog only provides technical information about the changes
          introduced with a particular Pico version, and is meant to supplement
          the actual code changes. The information in this changelog are often
          insufficient to understand the implications of larger changes. Please
          refer to both the UPGRADE and NEWS sections of the docs for more
          details.

**Note:** Changes breaking backwards compatibility (BC) are marked with an `!`
          (exclamation mark). This doesn't include changes for which BC is
          preserved by this plugin. If a previously deprecated feature is later
          removed, this change is going to be marked as BC-breaking change.
          Please note that BC-breaking changes are only possible with a new
          major version.

**Note:** Many versions of `PicoDeprecated` include changes which are not
          explicitly mentioned in this changelog. This primarily concerns
          changes in Pico's plugin API. These changes aren't listed separately
          because they are already listed in Pico's changelog. Only functional
          changes and/or BC-breaking changes are listed below.

### Version 3.0.0-beta.1
Released: -

```
* [New] Update to API version 4 (no functional changes necessary)
* [New] Add new continous integration (CI) pipeline using GitHub Actions
* [New] Add new build script and Makefile to simplify PicoDeprecated's build
        and release process; see Pico's `CONTRIBUTING.md` for details
* [New] Add `PicoDeprecated::VERSION` and `PicoDeprecated::VERSION_ID`
        constants (similar to Pico's matching constants)
* [New] Add plugin and theme compatibility plugin for API version 3
* [Changed] Enable PHP strict typing
* [Changed] ! Move classes to `\picocms\PicoDeprecated\…` PHP namespace
* [Fixed] Don't instantiate compatibility plugins multiple times
* [Fixed] Update Twig API usage to be compatible with Twig 3.3
```

### Version 3.0.0-alpha.2
Released: 2020-12-24

No changes

### Version 2.1.4
Released: 2020-08-29

No changes

### Version 2.1.3
Released: 2020-07-10

No changes

### Version 2.1.2
Released: 2020-04-10

No changes

### Version 3.0.0-alpha.1
Released: 2020-03-29

```
* [New] Kick-start development of Pico 3.0
```

### Version 2.1.1
Released: 2019-12-31

No changes

### Version 2.1.0
Released: 2019-11-24

No changes

### Version 2.1.0-beta.1
Released: 2019-11-03

```
* [New] Add support for the latest API v3 changes
* [New] Support disabled Twig autoescape prior to API v3
* [New] Re-introduce `theme_url` config variable
* [New] Re-introduce `prev_page`, `base_dir` and `theme_dir` Twig variables
* [New] Support loading additional plugins using API v1 `onPluginsLoaded` event
* [New] Re-introduce Pico v0.9 config constant `CACHE_DIR`
* [New] Add release & build system to test the plugin using PHP_CodeSniffer and
        to automatically create pre-built release packages
* [Changed] Split the plugin's functionality into multiple compatibility
            plugins (two for each API version, for plugins and themes resp.)
            and load the necessary compatibility plugins on demand only; also
            allow 3rd-party plugins to load their own compatibility plugins
```

### Version 2.0.5-beta.1
Released: 2019-01-03

```
* [New] Add `2.0.x-dev` alias for master branch to `composer.json`
```

### Version 2.0.4
Released: 2018-12-17

No changes

### Version 2.0.3
Released: 2018-12-03

No changes

### Version 2.0.2
Released: 2018-08-12

No changes

### Version 2.0.1
Released: 2018-07-29

No changes

### Version 2.0.0
Released: 2018-07-01

No changes

### Version 2.0.0-beta.3
Released: 2018-04-07

No changes

### Version 2.0.0-beta.2
Released: 2018-01-21

```
* [New] Add support for the latest API v2 changes
* [New] ! Add support for themes using the old `.html` file extension for Twig
        templates; however, starting with API v2 plugins might rely on `.twig`
        as file extension, making this a BC-breaking change regardless
```

### Version 2.0.0-beta.1
Released: 2017-11-05

**Note:** Pico's official `PicoDeprecated` plugin was moved to this separate
          repository in preparation for Pico 2.0. Refer to Pico's changelog for
          a list of changes to this plugin before Pico 2.0.

```
* [New] Update plugin to API v2 and add support for all API v1 events
* [New] Keep track of all loaded Pico plugins and distinguish them by the API
        version they use; deprecated events are only triggered on plugins using
        this particular API version (`PicoDeprecated::API_VERSION_*` constants)
* [New] Take care of triggering events on plugins using older API versions;
        this includes not only core events, but also all custom events; as a
        result, old plugin's always depend on `PicoDeprecated` now
* [New] Use a simple event alias table to keep track of unchanged or just
        renamed core events
* [New] Add `rewrite_url` and `is_front_page` Twig variables
* [New] Add support for the `config/config.php` configuration file
* [New] Additionally compare registered meta headers case-insensitive
* [New] Make meta headers on the first level of a page's meta data also
        available using a lowered key (as of Pico 1.0; i.e. `SomeKey: value` is
        now accessible using both `$meta['SomeKey']` and `$meta['somekey']`)
* [New] Add public `PicoDeprecated::triggersApiEvents()` method
* [New] Add public `PicoDeprecated::triggerEvent()` method (and the additional
        `$apiVersion` parameter) as replacement for the previously protected
        method of the same name
* [Fixed] ! Don't overwrite the global `$config` variable if it is defined
* [Fixed] ! Improve re-indexing of pages added by the API v0 event `get_pages`
* [Changed] No longer try to guess whether the plugin needs to be enabled or
            not, rather enable it by default (guessing was pretty error-prone)

PicoDeprecated.php
<?php
/**
 * This file is part of Pico. It's copyrighted by the contributors recorded
 * in the version control history of the file, available from the following
 * original location:
 *
 * <https://github.com/picocms/pico-deprecated/blob/master/PicoDeprecated.php>
 *
 * The file was previously part of the project's main repository; the version
 * control history of the original file applies accordingly, available from
 * the following original location:
 *
 * <https://github.com/picocms/Pico/blob/82a342ba445122182b898a2c1800f03c8d16f18c/plugins/00-PicoDeprecated.php>
 *
 * SPDX-License-Identifier: MIT
 * License-Filename: LICENSE
 */

declare(strict_types=1);

use picocms\PicoDeprecated\Plugin\MainPlugin;
use picocms\PicoDeprecated\PluginApiPluginInterface;
use picocms\PicoDeprecated\PluginInterface;

/**
 * Maintain backward compatibility to older Pico releases
 *
 * `PicoDeprecated`'s purpose is to maintain backward compatibility to older
 * versions of Pico, by re-introducing characteristics that were removed from
 * Pico's core.
 *
 * `PicoDeprecated` is basically a mandatory plugin for all Pico installs.
 * Without this plugin you can't use plugins which were written for other
 * API versions than the one of Pico's core, even when there was just the
 * slightest change.
 *
 * {@see https://picocms.org/plugins/deprecated/} for a full list of features.
 *
 * @author  Daniel Rudolf
 * @link    https://picocms.org
 * @license https://opensource.org/licenses/MIT The MIT License
 * @version 3.0
 */
class PicoDeprecated extends AbstractPicoPlugin
{
    /**
     * PicoDeprecated version
     *
     * @var string
     */
    public const VERSION = '3.0.0-dev';

    /**
     * PicoDeprecated version ID
     *
     * @var int
     */
    public const VERSION_ID = 30000;

    /**
     * API version used by this plugin
     *
     * @var int
     */
    public const API_VERSION = 4;

    /**
     * API version 0, used by Pico 0.9 and earlier
     *
     * @var int
     */
    public const API_VERSION_0 = 0;

    /**
     * API version 1, used by Pico 1.0
     *
     * @var int
     */
    public const API_VERSION_1 = 1;

    /**
     * API version 2, used by Pico 2.0
     *
     * @var int
     */
    public const API_VERSION_2 = 2;

    /**
     * API version 3, used by Pico 2.1
     *
     * @var int
     */
    public const API_VERSION_3 = 3;

    /**
     * API version 4, used by Pico 3.0
     *
     * @var int
     */
    public const API_VERSION_4 = 4;

    /**
     * Loaded plugins, indexed by API version
     *
     * @see PicoDeprecated::getPlugins()
     *
     * @var object[][]
     */
    protected $plugins = [];

    /**
     * Loaded compatibility plugins
     *
     * @see PicoDeprecated::getCompatPlugins()
     *
     * @var PluginInterface[]
     */
    protected $compatPlugins = [];

    /**
     * {@inheritDoc}
     */
    public function __construct(Pico $pico)
    {
        parent::__construct($pico);

        if (is_file(__DIR__ . '/vendor/autoload.php')) {
            require(__DIR__ . '/vendor/autoload.php');
        }

        if (!class_exists(MainPlugin::class)) {
            throw new RuntimeException(
                "Cannot find PicoDeprecated's 'vendor/autoload.php'. If you're using a Composer-based Pico install, "
                . "run `composer update`. If you're rather trying to use one of PicoDeprecated's pre-built release "
                . "packages, make sure to download PicoDeprecated's release package matching Pico's version named "
                . "'pico-deprecated-release-v*.tar.gz'; make sure not to download a source code package."
            );
        }

        if ($pico::VERSION_ID !== self::VERSION_ID) {
            throw new RuntimeException(
                'PicoDeprecated v' . self::VERSION . ' is incompatible with Pico v' . $pico::VERSION
            );
        }

        if ($pico::API_VERSION !== self::API_VERSION) {
            throw new RuntimeException(
                'PicoDeprecated requires API version ' . self::API_VERSION . ', '
                . 'but Pico is running API version ' . $pico::API_VERSION
            );
        }
    }

    /**
     * {@inheritDoc}
     */
    public function handleEvent($eventName, array $params)
    {
        parent::handleEvent($eventName, $params);

        // trigger events on compatibility plugins
        if ($this->isEnabled() || ($eventName === 'onPluginsLoaded')) {
            $isCoreEvent = in_array($eventName, $this->getCoreEvents());
            foreach ($this->compatPlugins as $plugin) {
                if ($isCoreEvent) {
                    if ($plugin->getApiVersion() === self::API_VERSION) {
                        $plugin->handleEvent($eventName, $params);
                    }
                } elseif ($plugin instanceof PluginApiPluginInterface) {
                    $plugin->handleCustomEvent($eventName, $params);
                }
            }
        }
    }

    /**
     * Reads all loaded plugins and indexes them by API level, loads the
     * necessary compatibility plugins
     *
     * @see PicoDeprecated::loadPlugin()
     *
     * @param object[] $plugins loaded plugin instances
     */
    public function onPluginsLoaded(array $plugins): void
    {
        $this->loadCompatPlugin(MainPlugin::class);

        foreach ($plugins as $plugin) {
            $this->loadPlugin($plugin);
        }

        $this->getPico()->triggerEvent('onPicoDeprecated', [ $this ]);
    }

    /**
     * Adds a manually loaded plugin to PicoDeprecated's plugin index, loads
     * the necessary compatibility plugins
     *
     * @see PicoDeprecated::loadPlugin()
     *
     * @param object $plugin loaded plugin instance
     */
    public function onPluginManuallyLoaded(object $plugin): void
    {
        $this->loadPlugin($plugin);
    }

    /**
     * Loads a compatibility plugin if Pico's theme uses a old theme API
     *
     * @param string $theme           name of current theme
     * @param int    $themeApiVersion API version of the theme
     * @param array  $themeConfig     config array of the theme
     */
    public function onThemeLoaded(string $theme, int $themeApiVersion, array &$themeConfig): void
    {
        $this->loadThemeApiCompatPlugin($themeApiVersion);
    }

    /**
     * Adds a plugin to PicoDeprecated's plugin index
     *
     * @see PicoDeprecated::onPluginsLoaded()
     * @see PicoDeprecated::onPluginManuallyLoaded()
     * @see PicoDeprecated::getPlugins()
     *
     * @param object $plugin loaded plugin instance
     */
    protected function loadPlugin(object $plugin): void
    {
        $pluginName = get_class($plugin);

        $apiVersion = $this->getPluginApiVersion($plugin);
        if (!isset($this->plugins[$apiVersion])) {
            $this->plugins[$apiVersion] = [];
            $this->loadPluginApiCompatPlugin($apiVersion);
        }

        $this->plugins[$apiVersion][$pluginName] = $plugin;
    }

    /**
     * Returns a list of all loaded Pico plugins using the given API level
     *
     * @param int $apiVersion API version to match plugins
     *
     * @return object[] loaded plugin instances
     */
    public function getPlugins(int $apiVersion): array
    {
        return $this->plugins[$apiVersion] ?? [];
    }

    /**
     * Loads a compatibility plugin
     *
     * @param PluginInterface|string $plugin either the class name of a plugin
     *     to instantiate or a plugin instance
     *
     * @return PluginInterface instance of the loaded plugin
     */
    public function loadCompatPlugin($plugin): PluginInterface
    {
        if (!is_object($plugin)) {
            $className = (string) $plugin;
            if (isset($this->compatPlugins[$className])) {
                return $this->compatPlugins[$className];
            }

            if (class_exists($className)) {
                $plugin = new $className($this->getPico(), $this);
            } else {
                throw new RuntimeException(
                    "Unable to load PicoDeprecated compatibility plugin '" . $className . "': Class not found"
                );
            }
        }

        $className = get_class($plugin);
        if (isset($this->compatPlugins[$className])) {
            return $this->compatPlugins[$className];
        }

        if (!($plugin instanceof PluginInterface)) {
            throw new RuntimeException(
                "Unable to load PicoDeprecated compatibility plugin '" . $className . "': "
                . "Compatibility plugins must implement '" . PluginInterface::class . "'"
            );
        }

        $apiVersion = $plugin->getApiVersion();
        $this->loadPluginApiCompatPlugin($apiVersion);

        $dependsOn = $plugin->getDependencies();
        foreach ($dependsOn as $pluginDependency) {
            $this->loadCompatPlugin($pluginDependency);
        }

        $this->compatPlugins[$className] = $plugin;

        return $plugin;
    }

    /**
     * Loads a plugin API compatibility plugin
     *
     * @param int $apiVersion API version to load the compatibility plugin for
     */
    protected function loadPluginApiCompatPlugin(int $apiVersion): void
    {
        if ($apiVersion !== self::API_VERSION) {
            $this->loadCompatPlugin('picocms\PicoDeprecated\Plugin\PluginApi' . $apiVersion . 'Plugin');
        }
    }

    /**
     * Loads a theme API compatibility plugin
     *
     * @param int $apiVersion API version to load the compatibility plugin for
     */
    protected function loadThemeApiCompatPlugin(int $apiVersion): void
    {
        if ($apiVersion !== self::API_VERSION) {
            $this->loadCompatPlugin('picocms\PicoDeprecated\Plugin\ThemeApi' . $apiVersion . 'Plugin');
        }
    }

    /**
     * Returns all loaded compatibility plugins
     *
     * @return PluginInterface[] list of loaded compatibility plugins
     */
    public function getCompatPlugins(): array
    {
        return $this->compatPlugins;
    }

    /**
     * Triggers deprecated events on plugins of different API versions
     *
     * You can use this public method in other plugins to trigger custom events
     * on plugins using a particular API version. If you want to trigger a
     * custom event on all plugins, no matter their API version (except for
     * plugins using API v0, which can't handle custom events), use
     * {@see Pico::triggerEvent()} instead.
     *
     * @see Pico::triggerEvent()
     *
     * @param int    $apiVersion API version of the event
     * @param string $eventName  event to trigger
     * @param array  $params     optional parameters to pass
     */
    public function triggerEvent(int $apiVersion, string $eventName, array $params = []): void
    {
        foreach ($this->getPlugins($apiVersion) as $plugin) {
            $plugin->handleEvent($eventName, $params);
        }
    }

    /**
     * Returns the API version of a given plugin
     *
     * @param object $plugin plugin instance
     *
     * @return int API version used by the plugin
     */
    public function getPluginApiVersion(object $plugin): int
    {
        $pluginApiVersion = self::API_VERSION_0;
        if ($plugin instanceof PicoPluginInterface) {
            $pluginApiVersion = self::API_VERSION_1;
            if (defined(get_class($plugin) . '::API_VERSION')) {
                $pluginApiVersion = $plugin::API_VERSION;
            }
        }

        return $pluginApiVersion;
    }

    /**
     * Returns a list of the names of Pico's core events
     *
     * @return string[] list of Pico's core events
     */
    public function getCoreEvents(): array
    {
        return [
            'onPluginsLoaded',
            'onPluginManuallyLoaded',
            'onConfigLoaded',
            'onThemeLoading',
            'onThemeLoaded',
            'onRequestUrl',
            'onRequestFile',
            'onContentLoading',
            'on404ContentLoading',
            'on404ContentLoaded',
            'onContentLoaded',
            'onMetaParsing',
            'onMetaParsed',
            'onContentParsing',
            'onContentPrepared',
            'onContentParsed',
            'onPagesLoading',
            'onSinglePageLoading',
            'onSinglePageContent',
            'onSinglePageLoaded',
            'onPagesDiscovered',
            'onPagesLoaded',
            'onCurrentPageDiscovered',
            'onPageTreeBuilt',
            'onPageRendering',
            'onPageRendered',
            'onMetaHeaders',
            'onYamlParserRegistered',
            'onParsedownRegistered',
            'onTwigRegistered',
        ];
    }
}
